crates/oxc_allocator/src/vec.rs

```diff
@@ -101,6 +101,33 @@ impl<'alloc, T> Vec<'alloc, T> {
         vec.extend(iter);
         Self(vec)
     }
+
+    /// Converts the vector into [`Box<[T]>`][owned slice].
+    ///
+    /// If the vector has excess capacity, its items will be moved into a
+    /// newly-allocated buffer with exactly the right capacity. That is, any
+    /// excess capacity will be removed.
+    ///
+    /// [owned slice]: Box
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use oxc_allocator::{Allocator, Vec};
+    ///
+    /// let allocator = Allocator::default();
+    /// let v = Vec::from_iter_in([1, 2, 3], &allocator);
+    /// let b = v.into_boxed_slice();
+    ///
+    /// assert_eq!(&*b, &[1, 2, 3]);
+    /// ```
+    pub fn into_boxed_slice(self) -> Box<'alloc, [T]> {
+        let b = self.0.into_boxed_slice();
+        let (ptr, _) = allocator_api2::boxed::Box::into_non_null(b);
+        // SAFETY: this is effectively a transmute from allocator_api2's box to
+        // our own representation. Ptr is non-null and points to initialized, well-aligned memory.
+        unsafe { Box::from_non_null(ptr) }
+    }
```

Comment:
Hmm. I'm not sure about this.

I'm not that familiar with `allocator_api2`'s implementation of `into_boxed_slice`, but it sounds like it can reallocate (copy all the data), which we definitely don't want.

That makes sense for `allocator_api2` as it's generic over any kind of allocator, but in our case we're using a bump allocator which cannot recover excess capacity, so it's pointless to copy the data.

I don't know if `bumpalo::Bump` would automatically ignore the request to reallocate or not. Would you like to look into that?

Alternatively, we could get the pointer and length of the `Vec` and construct a `NonNull<[T]>` manually. That's definitely do what we want. (Note: would need to wrap the `Vec` in a `ManuallyDrop` first, as `allocator_api2`'s `Vec` is `Drop`)

This all requires quite a bit of care. I'm not completely up to speed on the safety constraints. Sometimes there are non-obvious subtleties to avoiding UB, and I usually end up having to spend a few hours boning up on arcane details before I'm sure (well at least reasonably sure!) that I've not committed some UB crime!

Do you have a concrete use case for wanting a boxed slice to data in the arena?

Reply:
By the way, we're planning to replace this `Vec` implementation with a hand-written one with length and capacity fields as `u32`s, to bring size of `Vec` down from 32 bytes to 24 bytes (https://github.com/oxc-project/backlog/issues/18). Dunqing was going to look at doing that fairly soon.

So at that point, we wouldn't be able to use `allocator_api2`'s `into_boxed_slice` anyway.

Reply:
https://github.com/zakarumych/allocator-api2/blob/f937b9650e10137f782698bd4eb4f8b36f1817a7/src/stable/vec/mod.rs#L1046

I don't think it creates a new allocation. I want this because it cuts the size of a dynamically sized vec in half (32 -> 16 bytes) at the cost of not being re-sizeable.  I'm working on a prototype where this is useful.

Reply:
Actually, this works:

```rs
pub fn into_boxed_slice(self) -> Box<'alloc, [T]> {
    let slice = self.0.leak();
    let ptr = NonNull::from(slice);
    // SAFETY: `ptr` points to a valid slice `[T]`.
    // Lifetime of returned `Box<'alloc, [T]>` is same as lifetime of consumed `Vec<'alloc, T>`,
    // so data in the `Box` must be valid for its lifetime.
    unsafe { Box::from_non_null(ptr) }
}
```

Then you can remove the comment about "If the vector has excess capacity, its items will be moved into a newly-allocated buffer with exactly the right capacity."

Reply:
I've implemented this change in an extra commit.